src/envs/screws.py

```diff
@@ -0,0 +1,509 @@
+        # dimensions: [dx, dy, magnetized vs. unmagnetized].
+        lowers = np.array([
+            -(self.rz_x_ub - self.rz_x_lb), -(self.rz_y_ub - self.rz_y_lb), 0.0
+        ],
+                          dtype=np.float32)
+        uppers = np.array([(self.rz_x_ub - self.rz_x_lb),
+                           (self.rz_y_ub - self.rz_y_lb), 1.0],
+                          dtype=np.float32)
+        return Box(lowers, uppers)
+
+    def simulate(self, state: State, action: Action) -> State:
+        assert self.action_space.contains(action.arr)
+        # NOTE: currently, the agent can only pick things up or
+        # drop them iff its staying still. This isn't particularly
+        # realistic...
+        dx, dy, magnetization = action.arr
+        if dx != 0.0 or dy != 0.0:
+            return self._transition_move(state, dx, dy)
+
+        if magnetization > 0.5:
+            return self._transition_magnetize(state)
+        return self._transition_demagnetize(state)
+
+    def _get_tasks(self, num_tasks: int, possible_num_screws: List[int],
+                   rng: np.random.Generator) -> List[Task]:
+        tasks = []
+
+        # Sample center locations for each of the screw clusters.
+        screw_cluster_xcenters = []
+        for _ in range(self._num_screw_clusters):
+            screw_cluster_xcenters.append(
+                rng.uniform(self.rz_x_lb,
+                            self.rz_x_ub - self._receptacle_width))
+
+        for _ in range(num_tasks):
+            screw_name_to_pos: Dict[str, Tuple[float, float]] = {}
+            num_screws = rng.choice(possible_num_screws)
+            for si in range(num_screws):
+                existing_xys = set(screw_name_to_pos.values())
+                while True:
+                    # sample a random iniitial position for the screw
+                    # such that it is on the ground, not in the
+                    # receptacle, and near one of the screw clusters.
+                    y_pos = self.rz_y_lb + self._screw_height / 2.0
+                    screw_cluster_xcenter = rng.choice(screw_cluster_xcenters)
+                    x_pos = rng.normal(screw_cluster_xcenter)
+                    # if this position doesn't make this screw in collision
+                    # with another screw, then put the screw here.
+                    if self._surface_xy_is_valid(x_pos, y_pos, existing_xys):
+                        screw_name_to_pos[f"screw_{si}"] = (x_pos, y_pos)
+                        break
+            init_state, goal_atoms = \
+                self._get_init_state_and_goal_atoms_from_positions(
+                screw_name_to_pos, rng)
+            tasks.append(Task(init_state, goal_atoms))
+        return tasks
+
+    def _get_init_state_and_goal_atoms_from_positions(
+            self, screw_name_to_pos: Dict[str, Tuple[float, float]],
+            rng: np.random.Generator) -> Tuple[State, Set[GroundAtom]]:
+        data: Dict[Object, Array] = {}
```

Comment:
I'm recently a fan of using `utils.create_state_from_dict` instead for this, for improved readibility. see stick button or doors for examples.